### ISSUE TITLE
Check performance of initial document loading

### ISSUE THREAD

**Issue by rpominov** [webstudio-is/webstudio-designer#346]
Loading of server-rendered html documents like `https://alpha.webstudio.is/designer/{projectID}` takes from 300-700 ms currently.

- [ ] Check that code and database are in the same region
- [ ] Check what request we make to the database during server rendering and whether the number of request can be reduced
- [ ] Consider the possibility of loading all data on the client (no server render?)

**Comment by kof** [comment]
same with canvas/preview document

**Comment by kof** [comment]
I would still server-render, just without the data, server-render empty state basically .... we need to also think of what happens once we have a large site with lots of data

**Comment by rpominov** [comment]
TLDR: Regions are correct. If we remove all the data loading, best we can expect is to go from 480ms median load time to ~300ms for the initial document loading, and then we'll have to load data on the client.

Findings so far:

 - DB delays are fine. Looks like we're in the same region as expected
 - The median load time of designer (with my internet connection) is 480ms, which consists of:
   - 250ms network latency (that's how long it takes to load a [noop](https://github.com/webstudio-is/webstudio-designer/blob/c326fb1e503af4b16725bf3226f7140f1a69d093/apps/designer/app/routes/rest/noop.ts#L1-L3) request: `loader = () => {return { ok: 1 }}`)
   - about 30ms to server render, including DB requests (based on observing logs at Vercel https://vercel.com/webstudio-is/webstudio-designer/H3tbLQrBTsqDDQ5FtFCYYpiAqpyr/functions)
   - 130ms to download the html document (23kb)
   - 70ms not sure, I assume it's overhead of handling 23kb internally (sending between some buffers, compressing, etc)


--------------------

K6 Outputs:

https://alpha.webstudio.is/designer/14b5b20a-a315-41a1-ac07-f64914ca8dec

```
running (01m01.2s), 0/1 VUs, 100 complete and 0 interrupted iterations
default ✓ [======================================] 1 VUs  01m01.2s/10m0s  100/100 shared iters

     data_received..................: 8.7 MB 143 kB/s
     data_sent......................: 28 kB  459 B/s
     http_req_blocked...............: avg=2.83ms   min=0s       med=2µs      max=282.89ms p(90)=2µs      p(95)=3µs     
     http_req_connecting............: avg=794.86µs min=0s       med=0s       max=79.48ms  p(90)=0s       p(95)=0s      
     http_req_duration..............: avg=608.53ms min=315.86ms med=463.13ms max=2.23s    p(90)=995.38ms p(95)=1.36s   
       { expected_response:true }...: avg=608.53ms min=315.86ms med=463.13ms max=2.23s    p(90)=995.38ms p(95)=1.36s   
     http_req_failed................: 0.00%  ✓ 0        ✗ 100
     http_req_receiving.............: avg=64.07ms  min=9.03ms   med=35.16ms  max=698.95ms p(90)=103.57ms p(95)=147.38ms
     http_req_sending...............: avg=148.54µs min=65µs     med=141µs    max=761µs    p(90)=173.6µs  p(95)=187.29µs
     http_req_tls_handshaking.......: avg=1.14ms   min=0s       med=0s       max=114.53ms p(90)=0s       p(95)=0s      
     http_req_waiting...............: avg=544.3ms  min=294.22ms med=424.39ms max=1.53s    p(90)=889.06ms p(95)=1.23s   
     http_reqs......................: 100    1.634018/s
     iteration_duration.............: avg=611.92ms min=316.38ms med=463.66ms max=2.23s    p(90)=1.01s    p(95)=1.36s   
     iterations.....................: 100    1.634018/s
     vus............................: 1      min=1      max=1
     vus_max........................: 1      min=1      max=1
```


https://webstudio-designer-git-testnotformerge-webstudio-is.vercel.app/designer/9fcbbbd6-f684-4868-89f1-3a039b2c24cf

```
running (00m58.1s), 0/1 VUs, 100 complete and 0 interrupted iterations
default ✓ [======================================] 1 VUs  00m58.1s/10m0s  100/100 shared iters

     data_received..................: 9.1 MB 156 kB/s
     data_sent......................: 26 kB  451 B/s
     http_req_blocked...............: avg=4.22ms   min=1µs      med=2µs      max=422.04ms p(90)=2µs      p(95)=2µs     
     http_req_connecting............: avg=981.76µs min=0s       med=0s       max=98.17ms  p(90)=0s       p(95)=0s      
     http_req_duration..............: avg=575.82ms min=370.67ms med=479.35ms max=1.56s    p(90)=814.51ms p(95)=1.35s   
       { expected_response:true }...: avg=575.82ms min=370.67ms med=479.35ms max=1.56s    p(90)=814.51ms p(95)=1.35s   
     http_req_failed................: 0.00%  ✓ 0        ✗ 100
     http_req_receiving.............: avg=147.51ms min=884µs    med=131.86ms max=258.47ms p(90)=210.24ms p(95)=222.97ms
     http_req_sending...............: avg=194.86µs min=109µs    med=174.5µs  max=1.01ms   p(90)=226.8µs  p(95)=250.89µs
     http_req_tls_handshaking.......: avg=1.44ms   min=0s       med=0s       max=144.18ms p(90)=0s       p(95)=0s      
     http_req_waiting...............: avg=428.11ms min=269.29ms med=322.34ms max=1.35s    p(90)=649.25ms p(95)=1.18s   
     http_reqs......................: 100    1.721617/s
     iteration_duration.............: avg=580.79ms min=371.55ms med=480ms    max=1.56s    p(90)=847.03ms p(95)=1.35s   
     iterations.....................: 100    1.721617/s
     vus............................: 1      min=1      max=1
     vus_max........................: 1      min=1      max=1
```

https://webstudio-designer-git-testnotformerge-webstudio-is.vercel.app/rest/noop

```
running (00m28.1s), 0/1 VUs, 100 complete and 0 interrupted iterations
default ✓ [======================================] 1 VUs  00m28.1s/10m0s  100/100 shared iters

     data_received..................: 21 kB  742 B/s
     data_sent......................: 5.8 kB 207 B/s
     http_req_blocked...............: avg=2.07ms   min=0s       med=2µs      max=207.11ms p(90)=2µs      p(95)=2.04µs  
     http_req_connecting............: avg=923.85µs min=0s       med=0s       max=92.38ms  p(90)=0s       p(95)=0s      
     http_req_duration..............: avg=277.95ms min=249.36ms med=261.18ms max=862.96ms p(90)=278.68ms p(95)=326.68ms
       { expected_response:true }...: avg=277.95ms min=249.36ms med=261.18ms max=862.96ms p(90)=278.68ms p(95)=326.68ms
     http_req_failed................: 0.00%  ✓ 0        ✗ 100
     http_req_receiving.............: avg=245.7µs  min=64µs     med=141µs    max=4.04ms   p(90)=191.7µs  p(95)=234.74µs
     http_req_sending...............: avg=191.31µs min=89µs     med=196.5µs  max=258µs    p(90)=232.6µs  p(95)=246.05µs
     http_req_tls_handshaking.......: avg=1.13ms   min=0s       med=0s       max=113.17ms p(90)=0s       p(95)=0s      
     http_req_waiting...............: avg=277.51ms min=249.01ms med=260.43ms max=862.53ms p(90)=278.4ms  p(95)=326.35ms
     http_reqs......................: 100    3.561899/s
     iteration_duration.............: avg=280.81ms min=250.18ms med=261.91ms max=863.91ms p(90)=279.39ms p(95)=368.09ms
     iterations.....................: 100    3.561899/s
     vus............................: 1      min=1      max=1
     vus_max........................: 1      min=1      max=1
```

**Comment by rpominov** [comment]
We should probably start using Cloudflare proxy. This can make a big difference. It should remove network latency due to physical distance for all request that can be cached. Then we can start thinking in terms of how I can make this request easier to cache.

**Comment by kof** [comment]
cc @andrasbacsai what are your thoughts regarding cf proxy when it comes to designer? please create an issue if you think that's easily done.

**Comment by kof** [comment]
@rpominov I think your tests didn't include loading a large site data, which was my main concern with loading data upfront, you could try simulate that by scripting a site with thousands of instances

**Comment by rpominov** [comment]
Found out that if you make document large enough, there's an AWS limit https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-limits.html#function-configuration-deployment-and-execution

So if the response payload is > 6mb it'll crash with `LAMBDA_RUNTIME Failed to post handler success response. Http response code: 413.`

Something we need to fix somehow even if we keep everything as is for performance.

**Comment by kof** [comment]
@rpominov yeah the whole concept of inlining the data for perf falls apart when we are building an application, which is why applications usually load data separately and show a skeleton as quickly as possible to make it feel fast

**Comment by rpominov** [comment]
Yeah. So, when it doesn't crash the performance is not good either. With empty canvas the server start sending response after ~300ms, but when there's a medium size tree (410kb of JSON uncompressed + same in the form of server rendered html = 718kb total), it takes ~2.5s to start sending response.

Also Vercel doesn't show logs for some reason for these requests, not sure why yet.

Will be looking more into why there's such large increase depending on tree size. But probably it's not a good idea to server render tree + send it in JSON as well, indeed.

---------------------

Empty canvas: https://webstudio-designer-69fygn731-webstudio-is.vercel.app/empty?projectId=9fcbbbd6-f684-4868-89f1-3a039b2c24cf&mode=edit (3.1kb / 7.8kb uncompressed)

```
running (00m04.1s), 0/1 VUs, 10 complete and 0 interrupted iterations
default ✓ [======================================] 1 VUs  00m04.1s/10m0s  10/10 shared iters

     data_received..................: 85 kB  21 kB/s
     data_sent......................: 2.1 kB 516 B/s
     http_req_blocked...............: avg=31.73ms  min=1µs      med=2µs      max=317.32ms p(90)=31.73ms  p(95)=174.52ms
     http_req_connecting............: avg=5.23ms   min=0s       med=0s       max=52.36ms  p(90)=5.23ms   p(95)=28.8ms  
     http_req_duration..............: avg=372.68ms min=295.76ms med=327.22ms max=505.36ms p(90)=491.65ms p(95)=498.51ms
       { expected_response:true }...: avg=372.68ms min=295.76ms med=327.22ms max=505.36ms p(90)=491.65ms p(95)=498.51ms
     http_req_failed................: 0.00%  ✓ 0        ✗ 10 
     http_req_receiving.............: avg=14.14ms  min=196µs    med=388µs    max=115.12ms p(90)=23.09ms  p(95)=69.1ms  
     http_req_sending...............: avg=328.79µs min=202µs    med=244µs    max=990µs    p(90)=441µs    p(95)=715.49µs
     http_req_tls_handshaking.......: avg=13.4ms   min=0s       med=0s       max=134.06ms p(90)=13.4ms   p(95)=73.73ms 
     http_req_waiting...............: avg=358.21ms min=290.31ms med=326.59ms max=488.95ms p(90)=479.27ms p(95)=484.11ms
     http_reqs......................: 10     2.466318/s
     iteration_duration.............: avg=405.14ms min=296.37ms med=327.95ms max=808.44ms p(90)=536.3ms  p(95)=672.37ms
     iterations.....................: 10     2.466318/s
     vus............................: 1      min=1      max=1
     vus_max........................: 1      min=1      max=1
```

Medium size canvas: https://webstudio-designer-69fygn731-webstudio-is.vercel.app/less-big?projectId=9fcbbbd6-f684-4868-89f1-3a039b2c24cf&mode=edit (13.9kb / 718kb uncompressed)

```
running (00m32.0s), 0/1 VUs, 10 complete and 0 interrupted iterations
default ✓ [======================================] 1 VUs  00m32.0s/10m0s  10/10 shared iters

     data_received..................: 7.2 MB 225 kB/s
     data_sent......................: 17 kB  544 B/s
     http_req_blocked...............: avg=43.06ms  min=0s       med=1µs      max=430.6ms  p(90)=43.06ms p(95)=236.83ms
     http_req_connecting............: avg=5.49ms   min=0s       med=0s       max=54.91ms  p(90)=5.49ms  p(95)=30.2ms  
     http_req_duration..............: avg=3.15s    min=1.94s    med=2.47s    max=7.75s    p(90)=4.4s    p(95)=6.07s   
       { expected_response:true }...: avg=3.15s    min=1.94s    med=2.47s    max=7.75s    p(90)=4.4s    p(95)=6.07s   
     http_req_failed................: 0.00%  ✓ 0        ✗ 10 
     http_req_receiving.............: avg=977.81ms min=581.63ms med=886.03ms max=1.65s    p(90)=1.42s   p(95)=1.53s   
     http_req_sending...............: avg=444.69µs min=221µs    med=308µs    max=1.74ms   p(90)=545.9µs p(95)=1.14ms  
     http_req_tls_handshaking.......: avg=13.37ms  min=0s       med=0s       max=133.69ms p(90)=13.37ms p(95)=73.53ms 
     http_req_waiting...............: avg=2.17s    min=1.26s    med=1.5s     max=6.67s    p(90)=3.13s   p(95)=4.9s    
     http_reqs......................: 10     0.312429/s
     iteration_duration.............: avg=3.19s    min=1.94s    med=2.47s    max=8.18s    p(90)=4.44s   p(95)=6.31s   
     iterations.....................: 10     0.312429/s
     vus............................: 1      min=1      max=1
     vus_max........................: 1      min=1      max=1
```
